HPC-9034 Update Nx 16, React 18 and respective libraries

Onitoxan opened the pull request:
## Describe your changes:

- 🚀 Upgraded Nx to v16.3.1
- 🚀 Upgraded React to v18.2.0
- 🚀 Upgraded Jest to v29.5.0 (Nx upgrade related)
- 🚀 Upgraded React Router to v6.11.2
- 📦  Different package updates
- Updated `<Switch>` React Router v5 component to `<Routes>` React Router v6
- Updated `<Route>` structure, React Router v6 has been thought to try to use relative paths (Didn't want to refactor too much to not go out of the task, but we could define how to write it correctly and that is easy to understand)
- Added `usePrompt()` hook to show prompts in the application when we need
 
Please refer to the commit messages to check a more in-detailed explanation of the changes step by step

## How can I test the changes?:

You should init the app and test [here](http://localhost:4200)
To test the changes is important to:

1. Check the navigation through the page and check that no link has been broken
2. Check that in the Enketo form of any operation, you can't leave an unsaved form without a prompt appearing (For reload action and closing tab action, the browser prompt should appear instead of ours)
3. The application loads correctly
4. Check package.json to see if something seems off

## TO DO:
I would say there is nothing left to do of the task

## Does any part of this code spark :star: joy :star:?:

No special or remarkable things were done I would say. (Except becoming a full-time reader of documentation)

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on apps/hpc-cdm/src/app/components/enketo/index.tsx:
```diff
@@ -136,29 +132,16 @@ export const EnketoEditableForm = (props: Props) => {
-          event.returnValue = ''; // Chrome requires returnValue to be set.
-        }
-      };
+  /**
```
Don't comment out code, delete it.

Pl217 commented on apps/hpc-cdm/src/app/pages/operations-list.tsx:
```diff
@@ -13,7 +12,6 @@ interface Props {
 
 export default (props: Props) => {
   const loader = dataLoader([], getEnv().model.operations.getOperations);
-
```
No need to remove empty lines

Pl217 commented on apps/hpc-cdm/src/app/pages/operation.tsx:
```diff
@@ -50,20 +44,6 @@ const PageOperation = (props: Props) => {
               return (
                 <>
                   <PageMeta title={[operation.name]} />
-                  {/* <C.Toolbar>
```
Good! We don't want any dead code lying around

Pl217 reviewed (changes requested):
Please split the first commit where you upgrade from Nx 13 directly to Nx 15. Even though Nx migration scripts upgrade all at once, you don't have to commit all changes all at once. Please see this PR #287.

Also, splitting your changes into individual smaller commits helps me understand how these changes are made and makes the review tremendously easier.

Onitoxan committed:
HPC-8665: Upgrade to Nx 14

This commit is generated by running the following command:
`npx nx migrate 14.8.8`

This has changed `package.json` and upgraded bunch of Nx
packages, but also a few other. Manual changes were made
to either do bigger update or revert updates:
* Upgrade of all React libraries is reverted
  * react
  * react-dom
  * react-is
  * react-router-dom
  * @testing-library/react
  * @types/react
  * @types/react-dom
* `@types/node` upgraded to latest v18.16.16 instead of v18.7.18
* Downgrade of ESLing and related libraries stopped
  * eslint-plugin-jsx-a11y
  * eslint-plugin-react

Besides package updates, `nx migrate` command also
generated `migrations.json` file, which should not
be committed, but will be used as a blueprint for
migrations in the next step, when we run
`npx nx migrate --run-migrations`

Onitoxan committed:
HPC-8480: Upgrade to React 18

Changes to how React is attached to root DOM element
are done according to this migration guide
https://react.dev/blog/2022/03/08/react-18-upgrade-guide#updates-to-client-rendering-apis

Onitoxan committed:
➖ Remove `@mui/styles` deprecated library

`@mui/styles` is the JSS legacy styling solution.
They don't guarantee compatibility with React 18
https://github.com/mui/material-ui/issues/32142#issuecomment-1090002923

Realistically, this removal of `@mui/styles`
should have been done when we migrated from
MUI v4 to v5 in https://github.com/UN-OCHA/hpc-cdm/pull/316

In the official docs for the migration away from JSS, it is
recomended to replace its usages with `styled` or `sx` API,
with `sx` being recomended for overriding minor CSS
https://mui.com/system/getting-started/the-sx-prop/
https://mui.com/material-ui/migration/migrating-from-jss/

Onitoxan committed:
⬆️ HPC-9034: Upgrade `react-router-dom` to v6

React router v6 comes with a major refactor
on how this library is used
https://reactrouter.com/en/main/upgrading/v5

This commit only upgrades the library and doesn't
compile. Next few commits will migrate the code

Onitoxan committed:
HPC-9034: Replace all `<Switch>` elements to `<Routes>`

"React Router v6 introduces a `Routes` component
that is kind of like `Switch` In order to use v6,
you'll need to convert all your `<Switch>` elements to `<Routes>`"
https://reactrouter.com/en/main/upgrading/v5#upgrade-all-switch-elements-to-routes

Onitoxan committed:
HPC-9034: Replace `<Redirect to` with `<Navigate to`

`<Navigate to` is just a drop-in replacement in React router v6
https://reactrouter.com/en/main/upgrading/v5#use-usenavigate-instead-of-usehistory

Onitoxan committed:
HPC-9034: Remove usages of `useHistory()`

As per official React Router v6 migration docs,
we should use `useNavigate` instead of `useHistory`
https://reactrouter.com/en/main/upgrading/v5#use-usenavigate-instead-of-usehistory

Onitoxan committed:
HPC-8665: Upgrade to Nx 15

This commit is generated by running the following command:
`npx nx migrate 15.9.4`

This has changed `package.json` and upgraded bunch of Nx
packages, but also a few other. Manual changes were made
to either do bigger update or revert updates:
* All Jest packages are upgraded to latest v29:
  * `jest`
  * `babel-jest`
  * `ts-jest`
  * `@types/jest`
* `@types/express` upgraded to latest v4.17.17 instead of v4.17.14
* Downgrade of following libraries stopped
  * `eslint-plugin-react`
  * `react-router-dom`
  * `styled-components`
  * `@types/node`
  * `@types/react`
  * `@types/react-dom`

Besides package updates, `nx migrate` command also
generated `migrations.json` file, which should not
be committed, but will be used as a blueprint for
migrations in the next step, when we run
`npx nx migrate --run-migrations`

Onitoxan committed:
HPC-8665: Upgrade to Nx 16

This commit is generated by running the following command:
`npx nx migrate latest`

This has changed `package.json` and only manual change
is upgrading to latest Typescript version.

Besides package updates, `nx migrate` command also
generated `migrations.json` file, which should not
be committed, but will be used as a blueprint for
migrations in the next step, when we run
`npx nx migrate --run-migrations`

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 commented on package.json:
```diff
@@ -40,24 +40,24 @@
+    "@babel/preset-typescript": "7.16.7",
+    "@mui/material": "^5.9.1",
+    "@mui/styled-engine-sc": "^5.8.0",
+    "@mui/styles": "^5.9.1",
```
Downgrading Babel and MUI libraries is very likely due to bad conflict resolution when rebasing on top of #373. Amend this commit and don't touch these libraries, as they are irrelevant for Nx upgrade.

Pl217 commented on package.json:
```diff
@@ -58,16 +58,16 @@
-    "@testing-library/react": "12.1.5",
-    "@types/dotenv": "^8.2.0",
+    "@testing-library/react": "13.1.1",
+    "@types/dotenv": "^6.1.0",
```
Downgrade of `@types/dotenv` likely came from a bad rebase on top of #373. This package is not needed and will be removed in #376, but you should not touch it anyway.

Pl217 commented on package.json:
```diff
@@ -96,14 +96,14 @@
-    "react-dom": "17.0.2",
+    "lint-staged": "12.3.8",
+    "oidc-client": "^1.10.1",
+    "prettier": "2.6.2",
```
Don't downgrade `lint-staged`, `oidc-client` and `prettier`. The reason behind this downgrade seems to be a bad rebase once #373 got merged.

Pl217 commented on libs/hpc-ui/src/lib/i18n/index.tsx:
```diff
@@ -107,7 +107,7 @@ export class Translations<
         {cache.map((v, i) =>
           v.type === 'format' ? (
-            <span key={i}>{v.format.format(params)}</span>
+            <span key={i}>{v.format.format(params) as ReactNode}</span>
```
Type assertions are basically developers saying Typescript to shut up, because we think we know better than the TS compiler :) That is usually not the case, so type assertions should be avoided. Thus, lets remove `Date` as possible type  of `params` argument instead of doing this assertion here.

Pl217 commented on apps/hpc-ftsadmin/src/main.tsx:
```diff
@@ -1,11 +1,13 @@
-ReactDOM.render(
+const root = ReactDOM.createRoot(
+  document.getElementById('root') as HTMLElement
+);
```
In the other comment I am saying bad things about type assertion and how it should be avoided, because TS compiler usually knows better than us. Here, that's not the case, because we (the developers) know better than Typescript. DOM element with ID `'root'` definitely exists and `document.getElementById` should not return `null`.

But still, I like to avoid type assertions, so let's throw an error if root element cannot be found.
```ts
const rootElement = document.getElementById('root');
if (!rootElement) {
  throw new Error('Failed to find the root element');
}
const root = ReactDOM.createRoot(rootElement);
```

Pl217 commented on apps/hpc-cdm/src/main.tsx:
```diff
@@ -1,16 +1,18 @@
-ReactDOM.render(
+const root = ReactDOM.createRoot(
+  document.getElementById('root') as HTMLElement
+);
```
ditto

Pl217 commented on apps/hpc-cdm/src/app/components/enketo/powered-by-footer.tsx:
```diff
@@ -2,10 +2,18 @@ import React, { useContext } from 'react';
+      <use xlinkHref={KLogoRoute} />{' '}
+    </svg>
+  );
+};
```
Unfortunately, SVG logo doesn't load
![Screenshot_20230509_111146](https://user-images.githubusercontent.com/16909534/237051942-bd94c903-72a0-461d-a916-67fbf499bdb0.png)

Pl217 commented on apps/hpc-cdm/src/app/components/form-assignments-list.tsx:
```diff
@@ -8,7 +10,17 @@ import { t } from '../../i18n';
+  [`& .${classes.tooltip}`]: {
+    padding: 8,
+  },
+});
```
`StyledToolTip` is the unnecessary alias that you kept in this file (defined below). It was just defined to add some padding around the tooltip. The changes you made with all this context and everything just for a few pixels of padding is not worth it. I would just remove all the code related to padding.

Otherwise, if you think some padding is needed, use simple `Tooltip` (not `StyledToolTip`) and add padding using [the sx prop](https://mui.com/system/getting-started/the-sx-prop/):
```tsx
<Tooltip
  componentsProps={{
    tooltip: { sx: { padding: '8px' } },
  }}
  ...
```

Pl217 commented on apps/hpc-cdm/jest.config.ts:
```diff
@@ -1,4 +1,5 @@
-module.exports = {
+/* eslint-disable */
```
Why did you disable ESLint in all of these TS config files? I don't see any linting problem in these files.

Pl217 commented on jest.preset.js:
```diff
@@ -1,4 +1,4 @@
-const nxPreset = require('@nrwl/jest/preset');
+const nxPreset = require('@nrwl/jest/preset').default;
```
Why is this the only Jest config file kept in Javascript?

Onitoxan replied:
Nx migration didn't want to change it, I changed it by myself now

Pl217 replied:
There is a reason why Nx migration didn't change it. We cannot use `import ... from` syntax even if we make this a Typescript file. If we need to keep `require()`, then, let's leave it as JS. Sorry for misleading you.

Pl217 commented on package-lock.json:
```diff
@@ -22986,8 +22986,7 @@
       "integrity": "sha512-uveM3byMbthO+6tXZ1n2zm0W3uJCQYtwt/v5zV5I77v2v18u0ITkb8xwhsDD2i3V2Kye7SaNR6FFJ6lMuY/WqQ==",
-      "dev": true,
-      "requires": {}
+      "dev": true
```
There are plenty of changes on lockfile that you do in various commits, which just looks like a mistake. Since GitHub doesn't properly allow to comment on commits, all will be condensed in this comment:

- Changes to package lockfile shouldn't be made in commit 64d1e9198222d8c45414108dbc82cecf30f9b2d8.
- Entire next commit (0546baf8e061b01a793ea103ec6d741c8bc786b1) should be dropped. Package `eslint-plugin-react-hooks` has already been updated in #373. Just drop this commit during the rebase and install packages properly in commits that do some real updates.
- Commit after that (0f18afd75ae863b3d2143cc24d12129882dbfcaf) is having a duplicated commit message, but doesn't do anything it claims. Just drop it.
- Finally, third commit you need to drop is 48e7f8b3d6df0981aaf42509b495c3173303f012. Looks like you messed up a bit with git history, but it's easy to fix.

Pl217 commented on package.json:
```diff
@@ -47,36 +47,36 @@
+    "@typescript-eslint/eslint-plugin": "5.59.1",
+    "@typescript-eslint/parser": "5.59.1",
+    "babel-jest": "28.1.3",
+    "babel-plugin-styled-components": "2.0.7",
```
Downgrade of `babel-plugin-styled-components` looks like a bad conflict resolution after rebase on top of #373.

Pl217 commented on package.json:
```diff
@@ -89,25 +89,28 @@
-    "husky": "8.0.3",
+    "express": "4.18.2",
+    "fp-ts": "^2.7.1",
+    "husky": "7.0.4",
```
Downgrades of `fp-ts` and `husky` look like a bad conflict resolution after rebase on top of #373.

Pl217 commented on package.json:
```diff
@@ -89,25 +89,28 @@
-    "io-ts": "^2.2.20",
-    "jest": "27.5.1",
+    "intl-messageformat": "^9.4.6",
+    "io-ts": "^2.2.9",
```
Downgrades of `intl-messageformat` and `io-ts` look like a bad conflict resolution after rebase on top of #373.

Pl217 commented on package.json:
```diff
@@ -89,25 +89,28 @@
-    "prettier": "2.6.2",
-    "react": "18.1.0",
-    "react-dom": "18.1.0",
+    "prettier": "2.8.8",
```
You downgraded `prettier` in 7d3bbc1e3776ab929bdeabb42acc438db5bfdd35 and then back up in 9f2cb06fb85e935208b60b75b7da3f1abbc44145. Just don't touch it :)

Pl217 commented on package.json:
```diff
@@ -89,25 +89,28 @@
-    "react-dom": "18.1.0",
+    "prettier": "2.8.8",
+    "react": "18.2.0",
+    "react-dom": "18.2.0",
```
You should update React (and related packages) to latest v18 in 7d3bbc1e3776ab929bdeabb42acc438db5bfdd35 and not touch it again in 9f2cb06fb85e935208b60b75b7da3f1abbc44145

Pl217 commented on package.json:
```diff
@@ -89,25 +89,28 @@
-    "styled-components": "5.3.10",
-    "ts-jest": "27.1.4",
-    "typescript": "4.6.3"
+    "styled-components": "5.3.5",
```
Downgrade of `styled-components` looks like a bad conflict resolution after rebase on top of https://github.com/UN-OCHA/hpc-cdm/pull/373.

Pl217 commented on package.json:
```diff
@@ -89,25 +89,28 @@
+    "typescript": "4.8.4"
+  },
+  "dependencies": {
+    "core-js": "^3.30.1"
```
Why was `core-js` added as a dependency? Is it really needed? All other dependencies are specified as `devDependencies`, so should `core-js`.

Pl217 commented on package.json:
```diff
@@ -104,7 +104,7 @@
     "react-icons": "^4.3.1",
     "react-is": "18.2.0",
-    "react-router-dom": "5.3.0",
+    "react-router-dom": "6.4.1",
```
When updating, why don't you update to latest available version? v6.4.1 was released 8 months ago. Let's update to latest version available.

I will reuse this comment for another thing. After the commit where you update `react-router-dom`, there are three commits in which you do required migrations. You have divided work in areas where you make changes. That is good. There could be another approach, where you divide work in commits by relevant type of migration. For example, I see that you changed all `<Switch>` elements to `<Routes>`. The commit message could then point to the [exact section](https://reactrouter.com/en/main/upgrading/v5#upgrade-all-switch-elements-to-routes) in migration documentation.

But, you don't have to make this change, it's just how I would divide it. Maybe we should add some relevant excerpts from migration documentation to the commit message of 037ec60e917dbdd7550e3741513b8b06cdada3e6 and not have next three commits refer to this one as **previous commit**: "Please check the upgrading from v5 to v6 guide of the **previous commit** message to understand why we changing the structure"

Pl217 commented on package.json:
```diff
@@ -46,36 +46,36 @@
-    "@nrwl/web": "15.0.0",
-    "@nrwl/workspace": "15.0.0",
+    "@nrwl/js": "16.0.3",
+    "@nrwl/tao": "16.0.3",
```
Why was `@mui/styles` removed in 740e3d57d30afe8288dacbc00c961da58658b277? Also, `@nrwl/js` has been deprecated in favor of `@nx/js`, whereas `@nrwl/tao` has been deprecated in favor of `nx`.

Onitoxan replied:
https://github.com/mui/material-ui/issues/32142#issuecomment-1090002923

That github post is related on the decision of removing `@mui/styles` and related to the `@nrwl` packages remaning, the NX migration didn't change them automatically and when I tried to do it manually the Nx version of these two doesn't seem to be in npm (even though Nx in their documentation they speak of nx/tao and nx/js)

Pl217 commented on package.json:
```diff
@@ -46,36 +46,36 @@
+    "@nx/jest": "16.0.3",
+    "@nx/linter": "16.0.3",
+    "@nx/node": "16.0.3",
+    "@nx/web": "16.0.3",
```
Looks like you don't need `@nx/web` package, which you actually stopped using when you upgraded to Nx 15 in 9f2cb06fb85e935208b60b75b7da3f1abbc44145

Pl217 commented on package.json:
```diff
@@ -95,20 +95,22 @@
     "io-ts": "^2.2.9",
-    "jest": "28.1.3",
+    "jest": "29.4.3",
+    "jest-environment-jsdom": "~28.1.1",
```
Why do we need this `jest-environment-jsdom` package now? And why it isn't on v29 like rest of Jest packages?

Pl217 commented on jest.preset.js:
```diff
@@ -1,11 +1,21 @@
+  resolver: '@nx/jest/plugins/resolver',
   moduleFileExtensions: ['ts', 'js', 'html'],
   coverageReporters: ['html'],
+  /* TODO: Update to latest Jest snapshotFormat
+   * By default Nx has kept the older style of Jest Snapshot formats
+   * to prevent breaking of any existing tests with snapshots.
+   * It's recommend you update to the latest format.
+   * You can do this by removing snapshotFormat property
+   * and running tests with --update-snapshot flag.
+   * Example: "nx affected --targets=test --update-snapshot"
+   * More info: https://jestjs.io/docs/upgrading-to-jest29#snapshot-format
+   */
+  snapshotFormat: { escapeString: true, printBasicPrototype: true },
```
Use the new defaults for snapshot format, we don't have any snapshots in this repo anyway :) Just remove this config.

Pl217 commented on libs/hpc-core/.babelrc:
```diff
@@ -1,3 +1,5 @@
 {
-  "presets": ["@nrwl/web/babel"]
+  "presets": [
+    "@nrwl/js/babel"
```
Why did you keep `@nrwl/`? There are usages in other places too, so I won't comment everywhere. Search and replace them with `@nx/`

Onitoxan replied:
(commented in the prvious related issue about this)

Pl217 commented on apps/hpc-cdm/src/app/components/form-assignments-list.tsx:
```diff
@@ -96,9 +96,9 @@ const FormAssignmentsList = (props: Props) => {
-                          classes={{
-                            tooltip: classes.tooltip,
-                          }}
+                        classes={{
+                          tooltip: classes.tooltip,
+                        }}
```
In 5d4eef5583ac76359279e582fad41a74323b0ee1, you have changes to two `webpack.js` files which are the only ones belonging there. Rest of the changes are fixes of mistakes from previous commits. Don't commit those separately, but go back, rebase and amend (change) the relevant commits where changes like this formatting belong.

As a useful rule of thumb, any time you write "and" in your commit message, you are likely doing something that doesn't belong in that commit :)

Pl217 reviewed (changes requested):
Thanks for separating your work into commits! Means a ton while reviewing. I have reviewed your work by looking at commits and that's also how I commented. Don't know whether that will be obvious when I publish inline comments.

You still seem to need some git polishing, but you're getting there!

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Pl217 committed:
HPC-8665: Migrate project to Nx 14

This commit is the output of running command
`npx nx migrate --run-migrations`
Only manual changes are removals of `/* eslint-disable */`
directives from top of `jest.config.ts` files.
`ts-node` was automatically listed as a direct
dependency, because it is needed to read
Jest configuration files in TypeScript.

Full output of command run:

Ran 14-2-0-add-json-schema from nx
  Add JSON Schema to Nx configuration files

  UPDATE nx.json
  UPDATE workspace.json
---------------------------------------------------------
Ran 14-2-0-remove-default-collection from nx
  Remove default collection from configuration to switch to prompts for collection

  UPDATE nx.json
---------------------------------------------------------
Ran 14.3.4-create-target-defaults from nx
  Replace targetDependencies with targetDefaults

  UPDATE nx.json
---------------------------------------------------------
Ran update-jest-config-extensions from @nrwl/jest
  Update move jest config files to .ts files.

  DELETE jest.config.js
  CREATE jest.config.ts
  DELETE apps/hpc-cdm/jest.config.js
  CREATE apps/hpc-cdm/jest.config.ts
  UPDATE apps/hpc-cdm/tsconfig.app.json
  UPDATE apps/hpc-cdm/tsconfig.spec.json
  UPDATE workspace.json
  DELETE libs/hpc-core/jest.config.js
  CREATE libs/hpc-core/jest.config.ts
  UPDATE libs/hpc-core/tsconfig.lib.json
  UPDATE libs/hpc-core/tsconfig.spec.json
  DELETE libs/hpc-data/jest.config.js
  CREATE libs/hpc-data/jest.config.ts
  UPDATE libs/hpc-data/tsconfig.lib.json
  UPDATE libs/hpc-data/tsconfig.spec.json
  DELETE libs/hpc-dummy/jest.config.js
  CREATE libs/hpc-dummy/jest.config.ts
  UPDATE libs/hpc-dummy/tsconfig.lib.json
  UPDATE libs/hpc-dummy/tsconfig.spec.json
  DELETE apps/hpc-ftsadmin/jest.config.js
  CREATE apps/hpc-ftsadmin/jest.config.ts
  UPDATE apps/hpc-ftsadmin/tsconfig.app.json
  UPDATE apps/hpc-ftsadmin/tsconfig.spec.json
  DELETE libs/hpc-live/jest.config.js
  CREATE libs/hpc-live/jest.config.ts
  UPDATE libs/hpc-live/tsconfig.lib.json
  UPDATE libs/hpc-live/tsconfig.spec.json
  DELETE libs/hpc-ui/jest.config.js
  CREATE libs/hpc-ui/jest.config.ts
  UPDATE libs/hpc-ui/tsconfig.lib.json
  UPDATE libs/hpc-ui/tsconfig.spec.json
---------------------------------------------------------
Ran update-to-export-default from @nrwl/jest
  Update to export default in jest config and revert jest.preset.ts to jest.preset.js

  UPDATE jest.config.ts
  UPDATE jest.preset.js
  UPDATE apps/hpc-cdm/jest.config.ts
  UPDATE libs/hpc-core/jest.config.ts
  UPDATE libs/hpc-data/jest.config.ts
  UPDATE libs/hpc-dummy/jest.config.ts
  UPDATE apps/hpc-ftsadmin/jest.config.ts
  UPDATE libs/hpc-live/jest.config.ts
  UPDATE libs/hpc-ui/jest.config.ts
  UPDATE package.json
---------------------------------------------------------
Ran update-webpack-executor from @nrwl/web
  Update usages of webpack executors to @nrwl/webpack

  UPDATE workspace.json
---------------------------------------------------------

 >  NX   Running 'npm install' to make sure necessary packages are installed

Pl217 committed:
📌 Pin React packages to specific version

Like other React-related packages, use
exact versions for:
* react-icons
* react-helmet
* @types/react-helmet

Also, these packages are updated to latest version

Pl217 committed:
⚰️ Remove commented out code

Pl217 committed:
HPC-9034: Remove usages of `<Route exact>`

https://reactrouter.com/en/main/upgrading/v5#upgrade-all-switch-elements-to-routes
As official migration docs state:
`<Route exact>` is gone. Instead, routes with descendant
routes (defined in other components) use a trailing `*`
in their path to indicate they match deeply

Pl217 committed:
HPC-9034: Migrate to `<Route element>` in routes

All `<Route children>` from the React Router v5
have changed to `<Route element>` in v6. To migrate,
we move route element from the child position to a
named `element` prop.

https://reactrouter.com/en/main/upgrading/v5#advantages-of-route-element

Pl217 committed:
HPC-9034: Remove deprecated usages of `<Route render`

In React Router v6, the maintainers switched
away from React components to React elements.

Thus, hooks are now favored way of getting the
data, so we can no longer get route params
in methods of `<Route>`. Therefore, we need to
use `usaParams()` hook to get URL params and
parse them. In case of an error, plain JS `Error`
object is throw, so that `errorElement` defined
on the caller can define how to handle an error.

Pl217 committed:
HPC-9034: Use relative router paths

In React Router v6, paths inside `<Route path>`
became relative when `<Route>`s are nested.
In previous version, all paths were absolute

Co-authored-by: Onitoxan <anpesalado@gmail.com>

Pl217 committed:
HPC-8665: Migrate project to Nx 15

This commit is the output of running command
`npx nx migrate --run-migrations`
Manual changes:
* Removing legacy snapshot format from `jest.preset.js`,
which basically means not accepting the proposed changes
* Webpack config files for both apps are manually
adjusted to new Webpack config styles
* Installation of a ton of new packages is reverted:
  * @svgr/webpack
  * css-loader
  * style-loader
  * stylus
  * stylus-loader
  * url-loader
  * webpack (as a direct dependency)
  * webpack-merge

Full output of command run:

 >  NX   Running migrations from 'migrations.json'

>  NX   NOTE  Use "nx show projects" to read the list of projects.

  If you read the list of projects from workspace.json, use "nx show projects" instead.

Ran 15-7-0-split-configuration-into-project-json-files from @nrwl/workspace
  Split global configuration files into individual project.json files. This migration has been added automatically to the beginning of your migration set to retroactively make them work with the new version of Nx.

  CREATE apps/hpc-cdm/project.json
  CREATE apps/hpc-cdm-e2e/project.json
  CREATE libs/hpc-core/project.json
  CREATE libs/hpc-data/project.json
  CREATE libs/hpc-dummy/project.json
  CREATE apps/hpc-ftsadmin/project.json
  CREATE apps/hpc-ftsadmin-e2e/project.json
  CREATE libs/hpc-live/project.json
  CREATE libs/hpc-ui/project.json
  DELETE workspace.json
---------------------------------------------------------
Ran 15.0.0-migrate-to-inputs from nx
  Replace implicitDependencies with namedInputs + target inputs

  UPDATE nx.json
---------------------------------------------------------
Ran 15.0.0-prefix-outputs from nx
  Prefix outputs with {workspaceRoot}/{projectRoot} if needed

  UPDATE apps/hpc-cdm/project.json
  UPDATE apps/hpc-ftsadmin/project.json
  UPDATE libs/hpc-core/project.json
  UPDATE libs/hpc-data/project.json
  UPDATE libs/hpc-dummy/project.json
  UPDATE libs/hpc-live/project.json
  UPDATE libs/hpc-ui/project.json
---------------------------------------------------------
Ran add-babel-inputs from @nrwl/web
  Adds babel.config.json to the hash of all tasks

  UPDATE nx.json
---------------------------------------------------------
Ran update-babel-preset from @nrwl/web
  Update `@nrwl/web/babel` preset to `@nrwl/js/babel` for projects that have a .babelrc file.

  UPDATE package.json
  UPDATE libs/hpc-core/.babelrc
  UPDATE libs/hpc-data/.babelrc
  UPDATE libs/hpc-live/.babelrc
---------------------------------------------------------
Ran add-jest-inputs from @nrwl/jest
  Stop hashing jest spec files and config files for build targets and dependent tasks

  UPDATE nx.json
---------------------------------------------------------

>  NX  Jest Snapshot format changed in v29.
By default Nx kept the older style to prevent breaking of existing tests with snapshots.
It's recommend you update to the latest format.
You can do this in your project's jest config file.
Remove the snapshotFormat property and re-run tests with the --update-snapshot flag.
More info: https://jestjs.io/docs/upgrading-to-jest29#snapshot-format

Ran update-configs-jest-29 from @nrwl/jest
  Update jest configs to support jest 29 changes (https://jestjs.io/docs/upgrading-to-jest29)

  UPDATE libs/hpc-dummy/jest.config.ts
  UPDATE libs/hpc-core/jest.config.ts
  UPDATE libs/hpc-data/jest.config.ts
  UPDATE libs/hpc-live/jest.config.ts
  UPDATE jest.preset.js
---------------------------------------------------------
Ran install-webpack-rollup-dependencies from @nrwl/react
  Install new dependencies for React projects using Webpack or Rollup.

  UPDATE package.json
---------------------------------------------------------

          apps/hpc-cdm/webpack has been renamed to apps/hpc-cdm/webpack and a new apps/hpc-cdm/webpack
          has been created for your project hpc-cdm.
          You should consider inlining the logic from apps/hpc-cdm/webpack into apps/hpc-cdm/webpack.
          You can read our guide on how to do this here:

          https://nx.dev/packages/webpack/documents/webpack-config-setup

Ran react-webpack-config-setup from @nrwl/react
  Creates or updates webpack.config.js file with the new options for webpack.

  CREATE apps/hpc-cdm/webpack
  UPDATE apps/hpc-cdm/project.json
  CREATE apps/hpc-ftsadmin/webpack.config.js
  UPDATE apps/hpc-ftsadmin/project.json
---------------------------------------------------------
Ran add-eslint-inputs from @nrwl/linter
  Stop hashing eslint config files for build targets and dependent tasks

  UPDATE nx.json
---------------------------------------------------------
Ran add-eslint-ignore from @nrwl/linter
  Add node_modules to root eslint ignore

  CREATE .eslintignore
---------------------------------------------------------
Ran add-cypress-inputs from @nrwl/cypress
  Stop hashing cypress spec files and config files for build targets and dependent tasks

  UPDATE nx.json
---------------------------------------------------------

 >  NX   Running 'npm install' to make sure necessary packages are installed

Pl217 committed:
Replace `@nrwl/tao` with `nx`

`@nrwl/tao` has been deprecated in favor of `nx`

Pl217 committed:
HPC-8665: Migrate project to Nx 16

This commit is the output of running command
`npx nx migrate --run-migrations`
Commit contains full output of this command,
without any manual changes. Almost all of the
changes are rescoping from `@nrwl/` to `@nx/`
https://nx.dev/recipes/other/rescope
There are a few remaining `@nrwl/` references
in generators of `nx.json`, as they weren't
changed by migration tool, thus they are kept
as such, without manual change to `@nx/` or removal.

Full output of command run:

 >  NX   Running migrations from 'migrations.json'

Ran 16.0.0-remove-nrwl-cli from nx
  Remove @nrwl/cli.

  UPDATE package.json
---------------------------------------------------------
Ran 16.0.0-update-nx-cloud-runner from nx
  Replace @nrwl/nx-cloud with nx-cloud

  UPDATE nx.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/workspace
  Replace @nrwl/workspace with @nx/workspace

  UPDATE package.json
  UPDATE nx.json
  UPDATE migrations.json
---------------------------------------------------------
Ran 16-0-0-move-workspace-generators-into-local-plugin from @nrwl/workspace
  Generates a plugin called 'workspace-plugin' containing your workspace generators.

  DELETE tools/generators
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/js
  Replace @nrwl/js with @nx/js

  UPDATE package.json
  UPDATE libs/hpc-core/.babelrc
  UPDATE libs/hpc-data/.babelrc
  UPDATE libs/hpc-live/.babelrc
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/cypress
  Replace @nrwl/cypress with @nx/cypress

  UPDATE package.json
  UPDATE apps/hpc-cdm-e2e/project.json
  UPDATE apps/hpc-ftsadmin-e2e/project.json
  UPDATE nx.json
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/linter
  Replace @nrwl/linter with @nx/linter

  UPDATE package.json
  UPDATE apps/hpc-cdm/project.json
  UPDATE apps/hpc-cdm-e2e/project.json
  UPDATE apps/hpc-ftsadmin/project.json
  UPDATE apps/hpc-ftsadmin-e2e/project.json
  UPDATE libs/hpc-core/project.json
  UPDATE libs/hpc-data/project.json
  UPDATE libs/hpc-dummy/project.json
  UPDATE libs/hpc-live/project.json
  UPDATE libs/hpc-ui/project.json
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/eslint-plugin-nx
  Replace @nrwl/eslint-plugin-nx with @nx/eslint-plugin

  UPDATE package.json
  UPDATE migrations.json
  UPDATE .eslintrc.json
  UPDATE apps/hpc-cdm/.eslintrc.json
  UPDATE apps/hpc-ftsadmin/.eslintrc.json
  UPDATE libs/hpc-ui/.eslintrc.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/web
  Replace @nrwl/web with @nx/web

  UPDATE package.json
  UPDATE nx.json
  UPDATE apps/hpc-cdm/project.json
  UPDATE apps/hpc-cdm/webpack.js
  UPDATE apps/hpc-ftsadmin/project.json
  UPDATE apps/hpc-ftsadmin/webpack.config.js
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/react
  Replace @nrwl/react with @nx/react

  UPDATE package.json
  UPDATE nx.json
  UPDATE apps/hpc-cdm/.babelrc
  UPDATE apps/hpc-cdm/jest.config.ts
  UPDATE apps/hpc-cdm/tsconfig.app.json
  UPDATE apps/hpc-cdm/tsconfig.spec.json
  UPDATE apps/hpc-cdm/webpack.js
  UPDATE apps/hpc-ftsadmin/.babelrc
  UPDATE apps/hpc-ftsadmin/jest.config.ts
  UPDATE apps/hpc-ftsadmin/tsconfig.app.json
  UPDATE apps/hpc-ftsadmin/tsconfig.spec.json
  UPDATE apps/hpc-ftsadmin/webpack.config.js
  UPDATE apps/hpc-ftsadmin/webpack.js
  UPDATE libs/hpc-dummy/.babelrc
  UPDATE libs/hpc-ui/.babelrc
  UPDATE libs/hpc-ui/tsconfig.lib.json
  UPDATE libs/hpc-ui/tsconfig.spec.json
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/jest
  Replace @nrwl/jest with @nx/jest

  UPDATE package.json
  UPDATE apps/hpc-cdm/project.json
  UPDATE apps/hpc-ftsadmin/project.json
  UPDATE libs/hpc-core/project.json
  UPDATE libs/hpc-data/project.json
  UPDATE libs/hpc-dummy/project.json
  UPDATE libs/hpc-live/project.json
  UPDATE libs/hpc-ui/project.json
  UPDATE jest.config.ts
  UPDATE jest.preset.js
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/express
  Replace @nrwl/express with @nx/express

  UPDATE package.json
  UPDATE nx.json
  UPDATE migrations.json
---------------------------------------------------------
Ran update-16-0-0-add-nx-packages from @nrwl/node
  Replace @nrwl/node with @nx/node

  UPDATE package.json
  UPDATE nx.json
  UPDATE migrations.json
---------------------------------------------------------

 >  NX   Running 'npm install' to make sure necessary packages are installed

Co-authored-by: Onitoxan <anpesalado@gmail.com>

Pl217 committed:
Clarify MUI styled engine usage in webpack

This is mainly done to distinguish from
Enketo resolving section above

Pl217 committed:
Rename webpack config files

webpack.config.js is a standard name
of a webpack config file

Pl217 committed:
Enable HMR in CDM dev server

Since we already have HMR enabled in FTS Admin,
which requires installation of libraries:
* @pmmmwh/react-refresh-webpack-plugin and
* react-refresh
we could make use of HMR in CDM too.

Pl217 committed:
⬆️ Update `react-toastify` to `v9.1.3`

Pl217 committed:
Properly resolve linting error for useless fragment

Just suppressing the error in
db08b474e6dc3e2dbc7b52f47d9afb0784e765b8 was lazy

Pl217 committed:
✏️ Fix "Navitation" typo

Pl217 commented on package.json:
```diff
@@ -44,36 +44,36 @@
+    "@nx/node": "16.0.3",
+    "@nx/webpack": "16.0.3",
+    "@nx/workspace": "16.0.3",
+    "@nx/react": "16.0.3",
```
When I requested you to split your work and do gradual upgrades of Nx, you did split into individual commits, which is exactly what I wanted. But, each Nx upgrade comes with a migration tool, that updates `package.json` with newer versions of software, but also creates `migrations.json`, that you run with `nx migrate --run-migrations`. Exact steps are listed in [the guide](https://nx.dev/core-features/automate-updating-dependencies).

However, when you split updates into individual commits, you only made changes to `package.json` versions and didn't do migrations to make actual changes in our codebase. You should have upgraded gradually version-by-version, but also should have migrated in same manner.

Pl217 commented on apps/hpc-cdm/src/app/components/form-assignments-list.tsx:
```diff
@@ -8,7 +10,13 @@ import { t } from '../../i18n';
+  tooltip: `${PREFIX}-tooltip`,
+};
+
+const StyledAppContextConsumer = muiStyled(AppContext.Consumer)({});
```
There is absolutely no need for any context consumers and this CSS class, you just use simple `Tooltip` from MUI and style with `componentsProps` on it.

Pl217 commented on apps/hpc-cdm/src/app/components/enketo/powered-by-footer.tsx:
```diff
@@ -2,7 +2,7 @@ import React, { useContext } from 'react';
 import { styled } from '@unocha/hpc-ui';
 import logo from '../../../assets/logos/enketologo.png';
-import { ReactComponent as KLogo } from '../../../assets/logos/kobologo.svg';
+import KLogo from '../../../assets/logos/kobologo';
```
I don't know why you changed Enketo and KoBoToolbox logos at all. I have done same upgrades and didn't need to touch anything.

Pl217 commented on apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx:
```diff
@@ -20,71 +20,73 @@ interface Props {
+            <Route
+              path="/w/:windowId/*"
+              element={<ValidateParams {...props} lang={lang} />}
+            />
```
First big complaint regarding how you migrated to React Router v6 is the fact that you hardcoded the paths. There are bunch of places where you did that, so this comment is about all of them. The paths were previously defined in `apps/hpc-cdm/src/app/paths.ts` file and there is no need to hardcode them now.

Yes, resolving paths has changed. Previously they were absolute and there was `exact` option, but now they are relative to their parent path. You should have kept the centralized location where paths are defined.

We do have better readability with paths hardcoded, but it's terrible if something changes, since paths are scattered around. We can also improve readability with docstrings or with return type definitions.

Pl217 commented on apps/hpc-cdm/src/app/pages/operation-cluster-forms.tsx:
```diff
@@ -20,71 +20,73 @@ interface Props {
   );
 };
 
+interface ValidateParamsProps extends Props {
+  lang: LanguageKey;
+}
+const ValidateParams = (props: ValidateParamsProps) => {
+  const { operation, cluster, lang } = props;
+  const params = useParams();
+  const id = Number(params.windowId);
+  if (!isNaN(id)) {
+    const windows = operation.reportingWindows.filter((w) => w.id === id);
+    if (windows.length === 1) {
+      return (
+        <PageOperationClusterFormAssignments
+          window={windows[0]}
+          {...{ operation, cluster }}
+        />
+      );
+    }
+  }
+  return (
+    <>
+      <ClusterNavigation
+        operation={operation}
+        cluster={cluster}
+        showSettingsButton
+      />
+      <C.NotFound strings={t.get(lang, (s) => s.components.notFound)} />
+    </>
+  );
+};
```
Second big complaint regarding how you migrated to React Router v6 is that you have a bunch of patterns like this. We can no longer have functions inside `<Route>` elements where we could do pre-processing and verification of URL parameters, but you have done exact thing in all the places where this pattern occurs:

1. Add local "temporary" (you even named it `TemporalForm` in one place) component and process parameters with `useParams()` inside of it
2. Use this temporary component with `<React element>`, now that all elements should be passed via that prop

My biggest complaint is actually that you recognized this as a pattern and then copy pasted same structure all over the place. You named such construct a "temporal" (probably meant "temporary") component in one place - `TemporalForm`, but then seem to have realized why such construct is needed (to validate URL path params) and named it more descriptive `ValidateParams`. You have repeated the same pattern 5 times without moving this into a separate component!

From reading the documentation, the maintainers seem to want us to use `useParams()` in components directly, that is fine, but you shouldn't have introduced a temporary component. What you should have done, according to documentation is exactly what you did in `apps/hpc-cdm/src/app/pages/operation.tsx`. Just use `useParams()` is component directly, no temporary components needed. That way, instead of component's error being handled outside, component will be responsible for checking its own props.

But, like I said, we need to go one step further and have a component that will do this for us, in a generic manner.

Pl217 commented on apps/hpc-ftsadmin/project.json:
```diff
@@ -0,0 +1,106 @@
+        ],
+        "styles": [],
+        "scripts": [],
+        "webpackConfig": "@nx/react/plugins/webpack"
```
I realized here that our webpack config for FTS Admin isn't used at all. Not your fault, it is like that on `develop` now. It is defined like this in `workspace.json` currently, but it should be:
```suggestion
        "webpackConfig": "apps/hpc-ftsadmin/webpack.js"
```

Pl217 commented on package.json:
```diff
@@ -44,36 +44,36 @@
+    "@nx/workspace": "16.0.3",
+    "@nx/react": "16.0.3",
+    "@testing-library/react": "13.4.0",
+    "@types/dotenv": "^8.2.0",
```
You have fixed your previous bad conflict resolutions during a rebase, when a bunch of packages were reverting changes from #373. Now, you only have one bad conflict resolution, because you've brought back `@types/dotenv` which got removed in #376.

Pl217 reviewed (changes requested):
My main concerns are that you didn't do Nx migration in steps and that you haven't upgraded codebase to React Router v6 properly (we need generic component for validating params and centralized routes). More info inline.

Pl217 reviewed (approved):
Ok, I have pushed fixes that address all of my own comments from above. @Onitoxan I need you to take a look at my work and make sure I didn't break anything you've done. Especially routes, which were tricky.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review

Onitoxan commented:
I have reviewed commit per commit and everything works fine to merge, I just have some observations to know why you decided to do certain decisions and also some suggestions!:

In commit 53c9f508a667fab8d46606218cd5f8529f89ea36 : I think we could delete this commit, also I think it doesn't help transitioning to react v6 (Even if the documentation says so) Because we are nesting the components taking them out of the `component` attribute that actually seems to be more fitting and easy to transition to v6 `element` attribute. Also, you modify `operation.tsx` in this commit that doesn't seem to belong here. I would have updated this file in previous commit were you are deleting the comment and rename the commit message, I think it would make more sense 🤔 

In commit 7ca98cc743ba0d80c4ffb9f10a260f071669f6ff I see you modify `es.json` and some of the translations seems awkward (It is normal) would you want me to revise that file and update some typos?

In commit 3da19483a17bedf2162c9c0421766688755c12a7 in the file `paths.tsx` You let a comment on top of the code, why didn't you put it on the commit description instead in the code itself?
 
In commit 7b867044ff12ec264f0b1f63199c29d714b0d7dd I don't understand what's the use of HMR, if you could tell me, or link me a page to read about it I would thank it.


Despite these comments I don't have blocking change requests, it's ready for a merge from my side 👍

Pl217 reviewed (changes requested):
@Onitoxan, those are all valid concerns, thanks for your feedback. I am going to answer one-by-one.

> In commit [53c9f50](https://github.com/UN-OCHA/hpc-cdm/commit/53c9f508a667fab8d46606218cd5f8529f89ea36) : I think we could delete this commit, also I think it doesn't help transitioning to react v6 (Even if the documentation says so) Because we are nesting the components taking them out of the `component` attribute that actually seems to be more fitting and easy to transition to v6 `element` attribute. Also, you modify `operation.tsx` in this commit that doesn't seem to belong here. I would have updated this file in previous commit were you are deleting the comment and rename the commit message, I think it would make more sense thinking

You are completely right. This commit is there only because official docs are saying it's a smart thing to do when preparing to migrate from React Router v5 to v6. I was thinking about removing it myself, but since I got fooled by documentation and wasted my time on these useless changes, I decided to keep it to justify my time spent :smile: But, you're right, it is useless and I will get rid of it.

On the other hand, I don't agree that changes to `operation.tsx` should be moved to a commit where I remove commented out code. That commit is for cleanup and we should not mix concerns. Since I'm deleting commit and changes of `operation.tsx` need to be moved somewhere else anyway, its natural place is in same commit other components are also switched to use `useParams()` hook.

> In commit [7ca98cc](https://github.com/UN-OCHA/hpc-cdm/commit/7ca98cc743ba0d80c4ffb9f10a260f071669f6ff) I see you modify `es.json` and some of the translations seems awkward (It is normal) would you want me to revise that file and update some typos?

Sure, I forgot that we have a developer with powers greater than those of Google Translate - a native Spanish speaker :) I will use the messages which you shared with me on Slack.

> In commit [3da1948](https://github.com/UN-OCHA/hpc-cdm/commit/3da19483a17bedf2162c9c0421766688755c12a7) in the file `paths.tsx` You let a comment on top of the code, why didn't you put it on the commit description instead in the code itself?

Whereas you're partly right about this comment being slightly biased, meaning it reads well when viewed in a diff now, but would be slightly out of place once it becomes a legacy code.
However, I still want to convey the message why some routes must have leading/trailing slashes and why the code in that file is slightly weird. Thus, I will remove reference to React Router v5 in this comment. Also, since I want this message to be visible, it doesn't belong in commit message.

> In commit [7b86704](https://github.com/UN-OCHA/hpc-cdm/commit/7b867044ff12ec264f0b1f63199c29d714b0d7dd) I don't understand what's the use of HMR, if you could tell me, or link me a page to read about it I would thank it.

HMR stands for Hot Module Replacement (I actually thought R is for Reload), which is a [Webpack feature](https://webpack.js.org/concepts/hot-module-replacement/) that allows partial reloads of application.

> Despite these comments I don't have blocking change requests, it's ready for a merge from my side 👍

No, you've nailed a few very valid points here, so I will address them soon and then merge.

unocha-hpc reviewed (approved):
Checks have passed and this pull request is ready for manual review
